lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -0,0 +1,950 @@
+
+  // Forcibly derive conformance to CodingKey.
+  tc.checkConformancesInContext(enumDC, mutableEnumDC);
+
+  // "Synthesized" a null enum decl.
+  return {enumDecl, /* attempted to synthesize? */true};
+}
+
+// Creates a new var decl representing
+//
+//   let container : containerBase<keyType>
+//
+// containerBase is the name of the type to use as the base (either
+// KeyedEncodingContainer or KeyedDecodingContainer).
+static VarDecl*
+createKeyedContainer(ASTContext &C, DeclContext *DC, Identifier containerBase,
+                     Type keyType) {
+  // Look up Keyed*Container
+  auto keyedContainerDecl =
+    cast<NominalTypeDecl>(lookupFoundationTypeDecl(C, containerBase));
+
+  // Bind Keyed*Container to Keyed*Container<KeyType>
+  Type boundType[1] = { keyType };
+  auto containerType = BoundGenericType::get(keyedContainerDecl, Type(),
+                                             C.AllocateCopy(boundType));
+
+  // let container : Keyed*Container<KeyType>
+  auto containerDecl = new (C) VarDecl(/*IsStatic=*/false, /*IsLet=*/true,
+                                       /*IsCaptureList=*/false, SourceLoc(),
+                                       C.Id_container, containerType, DC);
+  containerDecl->setImplicit();
+  containerDecl->setInterfaceType(containerType);
+  return containerDecl;
+}
+
+static CallExpr*
+createContainerKeyedByCall(ASTContext &C, DeclContext *DC, Expr *base,
+                           Type returnType, NominalTypeDecl *param) {
+  // (keyedBy:)
+  auto keyedByDecl = new (C) ParamDecl(/*IsLet=*/true, SourceLoc(),
+                                       SourceLoc(), C.Id_keyedBy,
+                                       SourceLoc(), C.Id_keyedBy,
+                                       returnType, DC);
+  keyedByDecl->setImplicit();
+  keyedByDecl->setInterfaceType(returnType);
+
+  // container(keyedBy:) method name
+  auto paramList = ParameterList::createWithoutLoc(keyedByDecl);
+  DeclName callName(C, C.Id_container, paramList);
+
+  // base.container(keyedBy:) expr
+  auto unboundCall = new (C) UnresolvedDotExpr(base, SourceLoc(), callName,
+                                               DeclNameLoc(),
+                                               /*Implicit=*/true);
+
+  // CodingKeys.self expr
+  auto codingKeysExpr = new (C) DeclRefExpr(ConcreteDeclRef(param),
+                                            DeclNameLoc(), /*Implicit=*/true);
+  auto codingKeysMetaTypeExpr = new (C) DotSelfExpr(codingKeysExpr, SourceLoc(),
+                                                    SourceLoc());
+
+  // Full bound base.container(keyedBy: CodingKeys.self) call
+  Expr* args[1] = { codingKeysMetaTypeExpr };
+  Identifier argLabels[1] = { C.Id_keyedBy };
+  return CallExpr::createImplicit(C, unboundCall, C.AllocateCopy(args),
+                                  C.AllocateCopy(argLabels));
+}
+
+static void deriveBodyCodable_encode(AbstractFunctionDecl *encodeDecl) {
```

Comment:
Believe me, I know — I didn't enjoy propagating it. 😛  But figured that consistency was more important here than absolute correctness. I think we can do a pass on all of the derived conformances at some point to clean them all up in a way that's consistent. Some do a lot of stuff manually that was later pulled out into helper functions that were not backported, and stuff like that.